File: sentil-java/src/main/java/io/github/sedislab/sentil/Synthesis.java
package io.github.sedislab.sentil;

import java.util.List;
import java.util.function.ToDoubleFunction;

/** The smooth-robustness primitives, the synthesis numerics, and the black-box optimizers. */
public final class Synthesis {
    private Synthesis() {
    }

    /** The smooth, differentiable minimum of the values at the given temperature. */
    public static double softMin(double[] values, double temperature) {
        return NativeLib.softMin(values, temperature);
    }

    /** The smooth, differentiable maximum of the values at the given temperature. */
    public static double softMax(double[] values, double temperature) {
        return NativeLib.softMax(values, temperature);
    }

    /** Find an input sequence for the model that best satisfies the spec. */
    public static SynthesisResult synthesize(SystemModel model, Formula spec)
            throws SentilException {
        return synthesize(model, spec, null, null, Backend.AUTO, 0, 0);
    }

    /**
     * Find an input sequence for the model that best satisfies the spec. bounds and
     * smooth may be null, and maxIters and population of 0 take the defaults.
     */
    public static SynthesisResult synthesize(SystemModel model, Formula spec, Bounds bounds,
            SmoothConfig smooth, Backend backend, long maxIters, long population)
            throws SentilException {
        long boundsHandle = bounds == null ? 0L : bounds.handle();
        double temperature = smooth == null ? 0.0 : smooth.temperature;
        int kind = smooth == null ? 0 : smooth.kind.code();
        SynthesisResult result = NativeLib.synthesize(model.handle(), spec.handle(), boundsHandle,
                temperature, kind, smooth != null, maxIters, backend.code(), population);
        model.rethrowCallbackError();
        return result;
    }

    /** Minimize 1/2 u'Pu + q'u subject to Gu &lt;= h, with P symmetric positive-definite. */
    public static double[] solveQp(double[][] p, double[] q, double[][] g, double[] h, long maxIters)
            throws SentilException {
        int n = p.length;
        int m = g.length;
        return NativeLib.solveQp(Matrices.flatten(p, n), n, q, Matrices.flatten(g, n), m, h, maxIters);
    }

    /** Solve Ax = b for a symmetric positive-definite A. */
    public static double[] solveSpd(double[][] matrix, double[] rhs) throws SentilException {
        int n = matrix.length;
        return NativeLib.solveSpd(Matrices.flatten(matrix, n), n, rhs);
    }

    /** The eigendecomposition of a symmetric matrix. */
    public static EigenDecomposition symmetricEigen(double[][] matrix) throws SentilException {
        int n = matrix.length;
        double[] packed = NativeLib.symmetricEigen(Matrices.flatten(matrix, n), n);
        double[] values = new double[n];
        System.arraycopy(packed, 0, values, 0, n);
        double[][] vectors = new double[n][n];
        for (int i = 0; i < n; i++) {
            System.arraycopy(packed, n + i * n, vectors[i], 0, n);
        }
        return new EigenDecomposition(values, vectors);
    }

    private static Optimum optimum(double[] packed) {
        double[] point = new double[packed.length - 1];
        System.arraycopy(packed, 1, point, 0, point.length);
        return new Optimum(point, packed[0]);
    }

    /** Maximize a gradient objective from start, optionally inside bounds. */
    public static Optimum maximize(GradientObjective objective, double[] start, Bounds bounds,
            long maxIters) throws SentilException {
        if (bounds == null) {
            try (Bounds unbounded = Bounds.unbounded(start.length)) {
                return optimum(NativeLib.maximize(objective, start, unbounded.handle(), maxIters));
            }
        }
        return optimum(NativeLib.maximize(objective, start, bounds.handle(), maxIters));
    }

    /**
     * Maximize a scalar objective with gradient-free CMA-ES from start, optionally
     * inside bounds.
     */
    public static Optimum cmaEs(ToDoubleFunction<double[]> objective, double[] start, Bounds bounds,
            CmaConfig config) throws SentilException {
        if (bounds == null) {
            try (Bounds unbounded = Bounds.unbounded(start.length)) {
                return cmaEs(objective, start, unbounded, config);
            }
        }
        return optimum(NativeLib.cmaEs(objective, start, bounds.handle(), config.population,
                config.maxGenerations, config.initialStep, config.tolStep, config.seed));
    }

    /** Maximize a batch objective with CMA-ES, scoring a whole population at once. */
    public static Optimum cmaEsBatched(BatchObjective objective, double[] start, Bounds bounds,
            CmaConfig config) throws SentilException {
        if (bounds == null) {
            try (Bounds unbounded = Bounds.unbounded(start.length)) {
                return cmaEsBatched(objective, start, unbounded, config);
            }
        }
        return optimum(NativeLib.cmaEsBatched(objective, start, bounds.handle(), config.population,
                config.maxGenerations, config.initialStep, config.tolStep, config.seed));
    }

    /**
     * The tightest parameter in [lower, upper] for which the formula make builds holds
     * on every trace.
     */
    public static double mineTightestParameter(ParameterFormula make, List<Trace> traces,
            double lower, double upper) throws SentilException {
        long[] handles = new long[traces.size()];
        for (int i = 0; i < handles.length; i++) {
            handles[i] = traces.get(i).handle();
        }
        return NativeLib.mineTightestParameter(make, handles, lower, upper);
    }

    /**
     * Estimate a rare-event probability over a user-defined simulator by adaptive
     * multilevel splitting.
     */
    public static RareEventEstimate adaptiveMultilevelSplitting(AmsInterface simulator,
            long particles, double targetScore, long maxSteps, long seed) throws SentilException {
        return NativeLib.adaptiveMultilevelSplitting(simulator, particles, targetScore, maxSteps,
                seed);
    }
}

File: sentil-matlab/+sentil/Synthesis.m
classdef Synthesis
    %SYNTHESIS Open-loop trajectory synthesis and the smooth-robustness helpers.

    methods (Static)
        function r = synthesize(model, spec, bounds, backend, maxIters, population, smooth)
            %SYNTHESIZE Find an input sequence for the model that best satisfies the spec.
            if nargin < 3 || isempty(bounds)
                boundsHandle = uint64(0);
            else
                boundsHandle = bounds.Handle;
            end
            if nargin < 4, backend = sentil.Backend.Auto; end
            if nargin < 5, maxIters = 0; end
            if nargin < 6, population = 0; end
            if nargin < 7, smooth = sentil.SmoothConfig; end
            r = sentil_mex('synthesize', model.Handle, spec.Handle, boundsHandle, ...
                smooth.temperature, double(int32(smooth.kind)), double(maxIters), ...
                double(int32(backend)), double(population), model.Box);
            r.backend = sentil.Backend(r.backend);
        end

        function v = soft_min(values, temperature)
            %SOFT_MIN A smooth lower bound on the minimum of the values.
            if nargin < 2, temperature = 10; end
            v = sentil_mex('soft_min', double(values), double(temperature));
        end

        function v = soft_max(values, temperature)
            %SOFT_MAX A smooth upper bound on the maximum of the values.
            if nargin < 2, temperature = 10; end
            v = sentil_mex('soft_max', double(values), double(temperature));
        end

        function [point, value] = maximize(objective, start, bounds, maxIters)
            %MAXIMIZE Climb an objective returning [value, gradient] from start within bounds.
            if nargin < 3 || isempty(bounds)
                bounds = sentil.Bounds.unbounded(numel(start));
            end
            if nargin < 4, maxIters = 0; end
            [point, value] = sentil_mex('maximize', objective, double(start), bounds.Handle, ...
                double(maxIters));
        end

        function [point, value] = cma_es(objective, start, bounds, config)
            %CMA_ES Maximize a scalar objective with gradient-free CMA-ES.
            if nargin < 3 || isempty(bounds)
                bounds = sentil.Bounds.unbounded(numel(start));
            end
            if nargin < 4, config = sentil.CmaConfig; end
            [point, value] = sentil_mex('cma_es', objective, double(start), bounds.Handle, ...
                config.population, config.max_generations, config.initial_step, config.tol_step, ...
                config.seed);
        end

        function [point, value] = cma_es_batched(objective, start, bounds, config)
            %CMA_ES_BATCHED CMA-ES with an objective that scores a whole population at once.
            if nargin < 3 || isempty(bounds)
                bounds = sentil.Bounds.unbounded(numel(start));
            end
            if nargin < 4, config = sentil.CmaConfig; end
            [point, value] = sentil_mex('cma_es_batched', objective, double(start), ...
                bounds.Handle, config.population, config.max_generations, config.initial_step, ...
                config.tol_step, config.seed);
        end

        function p = mine_tightest_parameter(make, traces, lower, upper)
            %MINE_TIGHTEST_PARAMETER The tightest parameter in [lower, upper] for which make(param) holds on every trace.
            handles = zeros(1, numel(traces), 'uint64');
            for i = 1:numel(traces)
                handles(i) = traces(i).Handle;
            end
            p = sentil_mex('mine_tightest_parameter', make, handles, double(lower), double(upper));
        end
    end
end